File: Engine/src/tungsten/frontend.py
from __future__ import annotations

from pathlib import Path

from .docs_index import DocumentationIndex
from .kernel import KernelEvaluationResult, WolframKernelRunner
from .notebook import wl_string


class FrontEndController:
    def __init__(
        self,
        runner: WolframKernelRunner | None = None,
        docs_index: DocumentationIndex | None = None,
    ) -> None:
        self.runner = runner or WolframKernelRunner()
        self.docs_index = docs_index or DocumentationIndex(self.runner.installation)

    def probe(self) -> dict[str, object]:
        result = self.runner.evaluate_text(
            'nb = UsingFrontEnd[CreateDocument[Notebook[{Cell["Tungsten probe", "Text"]}, Visible -> False]]];'
            " head = Head[nb];"
            " UsingFrontEnd[NotebookClose[nb]];"
            " head"
        )
        return result.to_dict()

    def run(self, code: str, *, wrap_using_front_end: bool = True) -> KernelEvaluationResult:
        return self.runner.evaluate_text(code, require_front_end=wrap_using_front_end)

    def open_notebook(self, path: Path) -> KernelEvaluationResult:
        return self.runner.evaluate_text(
            f"NotebookOpen[{wl_string(str(path.resolve().as_posix()))}]",
            require_front_end=True,
        )

    def open_documentation(
        self,
        identifier: str,
        *,
        index_path: Path | None = None,
    ) -> KernelEvaluationResult:
        paclet = self.docs_index.resolve_identifier(identifier, index_path=index_path)
        return self.runner.evaluate_text(
            f"NotebookLocate[{wl_string(paclet)}]",
            require_front_end=True,
        )

    def execute_token(
        self,
        token: str,
        *,
        notebook_path: Path | None = None,
    ) -> KernelEvaluationResult:
        if notebook_path is None:
            code = f"FrontEndTokenExecute[{wl_string(token)}]"
        else:
            notebook_literal = wl_string(str(notebook_path.resolve().as_posix()))
            code = (
                f"nb = NotebookOpen[{notebook_literal}];"
                f" FrontEndTokenExecute[nb, {wl_string(token)}];"
                " nb"
            )
        return self.runner.evaluate_text(code, require_front_end=True)

File: Engine/haskell/src/Tungsten/Frontend.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Narrow, typed FrontEnd operations over the structured kernel runner.
module Tungsten.Frontend
  ( probeFrontEnd
  , runFrontEnd
  , openNotebook
  , openDocumentation
  , executeFrontEndToken
  , frontEndProbeCode
  , openNotebookCode
  , openDocumentationCode
  , frontEndTokenCode
  ) where

import Data.Text (Text)
import qualified Data.Text as T
import System.Directory (makeAbsolute)
import Tungsten.Discovery
import Tungsten.Expression (Expr (String), fullForm)
import Tungsten.Kernel

probeFrontEnd :: WolframInstallation -> IO KernelEvaluationResult
probeFrontEnd installation =
  evaluateKernelText installation frontEndProbeCode Nothing False

runFrontEnd :: WolframInstallation -> Text -> Bool -> IO KernelEvaluationResult
runFrontEnd installation code wrapUsingFrontEnd =
  evaluateKernelText installation code Nothing wrapUsingFrontEnd

openNotebook :: WolframInstallation -> FilePath -> IO KernelEvaluationResult
openNotebook installation path = do
  absolutePath <- makeAbsolute path
  evaluateKernelText installation (openNotebookCode absolutePath) Nothing True

openDocumentation :: WolframInstallation -> Text -> IO KernelEvaluationResult
openDocumentation installation identifier =
  evaluateKernelText installation (openDocumentationCode identifier) Nothing True

executeFrontEndToken
  :: WolframInstallation
  -> Text
  -> Maybe FilePath
  -> IO KernelEvaluationResult
executeFrontEndToken installation token notebookPath = do
  absolutePath <- traverse makeAbsolute notebookPath
  evaluateKernelText installation (frontEndTokenCode token absolutePath) Nothing True

frontEndProbeCode :: Text
frontEndProbeCode =
  "nb = UsingFrontEnd[CreateDocument[Notebook[{Cell[\"Tungsten probe\", \"Text\"]}, Visible -> False]]];"
    <> " head = Head[nb];"
    <> " UsingFrontEnd[NotebookClose[nb]];"
    <> " head"

openNotebookCode :: FilePath -> Text
openNotebookCode path = "NotebookOpen[" <> stringLiteral (T.pack (normalizePath path)) <> "]"

openDocumentationCode :: Text -> Text
openDocumentationCode identifier = "NotebookLocate[" <> stringLiteral identifier <> "]"

frontEndTokenCode :: Text -> Maybe FilePath -> Text
frontEndTokenCode token Nothing =
  "FrontEndTokenExecute[" <> stringLiteral token <> "]"
frontEndTokenCode token (Just notebookPath) =
  "nb = NotebookOpen["
    <> stringLiteral (T.pack (normalizePath notebookPath))
    <> "]; FrontEndTokenExecute[nb, "
    <> stringLiteral token
    <> "]; nb"

stringLiteral :: Text -> Text
stringLiteral = fullForm . String

normalizePath :: FilePath -> FilePath
normalizePath = map (\character -> if character == '\\' then '/' else character)
